velox/exec/fuzzer/CacheFuzzer.cpp

```diff
@@ -0,0 +1,358 @@
+    }
+  }
+}
+
+void CacheFuzzer::initializeCache() {
+  // We have up to 20 threads and 16 threads are used for reading so
+  // there are some threads left over for SSD background write.
+  executor_ = std::make_unique<folly::IOThreadPoolExecutor>(20);
+
+  std::unique_ptr<SsdCache> ssdCache;
+  if (FLAGS_ssd_cache_bytes > 0) {
+    SsdCache::Config config(
+        fmt::format("{}/cache", sourceDataDir_->getPath()),
+        FLAGS_ssd_cache_bytes,
+        FLAGS_num_ssd_cache_shards,
+        executor_.get(),
+        FLAGS_ssd_checkpoint_interval_bytes,
+        false,
+        FLAGS_enable_checksum,
+        FLAGS_enable_checksum_read_verification);
+    ssdCache = std::make_unique<SsdCache>(config);
+  }
+
+  memory::MemoryManagerOptions options;
+  options.useMmapAllocator = true;
+  options.allocatorCapacity = FLAGS_memory_cache_bytes;
+  options.arbitratorCapacity = FLAGS_memory_cache_bytes;
+  options.arbitratorReservedCapacity = 0;
+  options.trackDefaultUsage = true;
+  memoryManager_ = std::make_unique<memory::MemoryManager>(options);
+  allocator_ =
+      dynamic_cast<memory::MmapAllocator*>(memoryManager_->allocator());
+
+  // TODO: Test different ssd write behaviors with AsyncDataCache::Options.
+  cache_ = AsyncDataCache::create(allocator_, std::move(ssdCache), {});
+}
+
+void CacheFuzzer::initializeInputs() {
+  auto readOptions = io::ReaderOptions(pool_.get());
+  auto tracker = std::make_shared<ScanTracker>(
+      "testTracker", nullptr, io::ReaderOptions::kDefaultLoadQuantum);
+  auto ioStats = std::make_shared<IoStatistics>();
+  inputs_.reserve(FLAGS_num_source_files);
+  for (auto i = 0; i < FLAGS_num_source_files; ++i) {
+    // Initialize buffered input.
+    auto readFile = fs_->openFileForRead(fileNames_[i]);
+    groupIds_.emplace_back(
+        fileIds(), fmt::format("group{}", i / FLAGS_num_files_in_group));
+    auto const withExecutor =
+        boost::random::uniform_int_distribution<int8_t>(0, 1)(rng_) == 1;
+    inputs_.emplace_back(std::make_unique<CachedBufferedInput>(
+        std::move(readFile),
+        MetricsLog::voidLog(),
+        fileIds_[i].id(), // NOLINT
+        cache_.get(),
+        tracker,
+        groupIds_[i].id(),
+        ioStats,
+        withExecutor ? executor_.get() : nullptr,
+        readOptions));
+
+    // Divide file into fragments.
+    std::vector<std::pair<int32_t, int32_t>> fragments;
+    int32_t offset = 0;
+    while (offset < fileSizes_[i]) {
+      const auto length = boost::random::uniform_int_distribution<int32_t>(
+          1, fileSizes_[i] - offset)(rng_);
+      fragments.emplace_back(offset, length);
+      offset += length;
+    }
+    fileFragments_.emplace_back(std::move(fragments));
+  }
+}
+
+void CacheFuzzer::readCache() {
+  runThreads(FLAGS_num_threads, [&](int32_t /*i*/) {
+    auto numReads = boost::random::uniform_int_distribution<int32_t>(
+        10, FLAGS_max_num_reads)(rng_);
+    while (numReads-- > 0) {
```

Comment:
Let's control the test iteration by time instead of the number of reads? Thanks!